lychee-lib/src/extract/markdown.rs

```diff
@@ -77,6 +82,93 @@ pub(crate) fn extract_markdown(input: &str, include_verbatim: bool) -> Vec<RawUr
         .collect()
 }
 
+/// Extract unparsed URL strings from a Markdown string.
+pub(crate) fn extract_markdown_fragments(input: &str, frag_t: FragmentGen) -> Vec<String> {
+    let mut in_heading = false;
+    let mut heading = String::new();
+    let mut id_counter = HashMap::<String, usize>::new();
+
+    Parser::new(input)
+        .filter_map(|event| match event {
+            Event::Start(Tag::Heading(..)) => {
+                in_heading = true;
+                None
+            }
+            Event::End(Tag::Heading(_d, id, ..)) => {
+                let frag = match &frag_t {
+                    FragmentGen::HeadingAttribute => id.map(ToString::to_string),
+                    FragmentGen::KebabCase => Some(into_kebab_case(&heading)),
+                    FragmentGen::UniqueKebabCase => {
+                        let mut id = into_kebab_case(&heading);
+
+                        let id_count = id_counter.entry(id.clone()).or_insert(0);
+                        if *id_count != 0 {
+                            id = format!("{id}-{id_count}");
+                        }
+                        *id_count += 1;
+                        Some(id)
+                    }
+                };
+                in_heading = false;
+                heading.clear();
+                frag
+            }
+            Event::Text(text) => {
+                if in_heading {
+                    heading.push_str(&text);
+                };
+                None
+            }
+
+            // Silently skip over other events
+            _ => None,
+        })
+        .collect()
+}
+
+/// Describes how fragments should be generated
+/// from a markdown document.
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq)]
+pub enum FragmentGen {
+    /// Only take fragments specified in heading attributes.
+    HeadingAttribute,
+    /// Convert Headings into Kebab Case for use as ids.
+    KebabCase,
+    /// KebabCase but append a number if duplicates exist.
+    UniqueKebabCase,
+}
```

Comment:
`UniqueKebabCase` currently ignores heading attributes, but I agree using heading attributes with a fallback to unique kebab case would make a great default.

Also note, fragments don't appear to be covered in the [CommonMark spec](https://spec.commonmark.org/0.30/). The unique kebab case method appears to be used by mdBook because it is used in GitHub.